ops/model.py

```diff
@@ -741,7 +752,7 @@ def bind_address(self) -> Optional['_IPAddress']:
             return None
 
     @property
-    def ingress_address(self):
+    def ingress_address(self) -> Optional['_NetworkAddress']:
```

Comment:
Since charmers will likely have only ever written code that expects this to return an ip address object (not a str), they already have broken code.  But they probably have written code that handles this returning None.  So would it be safer for us to log maybe an error - and then return none from here if the cast fails?  This basically would divert the error case down code paths charms already have established and not introduce any new error modes.  The danger is that there _is_ some charm that can handle a string and wants it despite not being able to resolve it to an ip.

Reply:
> Since charmers will likely have only ever written code that expects this to return an ip address object (not a str), they already have broken code. But they probably have written code that handles this returning None. So would it be safer for us to log maybe an error - and then return none from here if the cast fails? This basically would divert the error case down code paths charms already have established and not introduce any new error modes. The danger is that there _is_ some charm that can handle a string and wants it despite not being able to resolve it to an ip.

IPs and hostnames are pretty much equivalents of each other in a lot of circumstances. They might lead to branching codepaths once you get to the level of the OS's networking stack. But if I just want to make an http call to an API, my code doesn't need to care about whether or not the string representing the API's address is an IP or a hostname.

In other words, I think that we do risk breaking working charms by being strict about it.

Or to put it another way: the object isn't actually an IP address. It's the host part of the URI, and that can be a set or numbers _or_ a string of letters.

Reply:
I mostly agree with that.  But will charm code have been written treating the return object explicitly like an ip_address object?  If so, then returning a string very well could break charm code anyway.  But maybe that's just a pill we need to swallow here.

Reply:
It feels most "correct" to me to accept either an ip address or a hostname. I'm not sure what choices charmers have made.

But I don't think that this change breaks them any more than they'd already be broken, if they are breaking on things that aren't ip addresses.

Reply:
Okay - I'm convinced that this isn't more bad than the alternatives :P